pkg/server/render.go

```diff
@@ -106,3 +129,53 @@ func (ctrl *Controller) renderParametersFromRequest(r *http.Request, p *renderPa
 	p.format = v.Get("format")
 	return nil
 }
+
+func parseRenderRangeParams(v url.Values, from, until string) (startTime, endTime time.Time, ok bool) {
+	fromStr, untilStr := v.Get(from), v.Get(until)
+	startTime, endTime = attime.Parse(fromStr), attime.Parse(untilStr)
+	return startTime, endTime, fromStr != "" || untilStr != ""
+}
+
+func (ctrl *Controller) loadTreeConcurrently(
+	gi *storage.GetInput,
+	treeStartTime, treeEndTime time.Time,
+	leftStartTime, leftEndTime time.Time,
+	rghtStartTime, rghtEndTime time.Time,
+) (treeOut, leftOut, rghtOut *storage.GetOutput, _ error) {
+
+	var treeErr, leftErr, rghtErr error
+	var wg sync.WaitGroup
+	wg.Add(3)
+	go func() { defer wg.Done(); treeOut, treeErr = ctrl.loadTree(gi, treeStartTime, treeEndTime) }()
+	go func() { defer wg.Done(); leftOut, leftErr = ctrl.loadTree(gi, leftStartTime, leftEndTime) }()
+	go func() { defer wg.Done(); rghtOut, rghtErr = ctrl.loadTree(gi, rghtStartTime, rghtEndTime) }()
+	wg.Wait()
+
+	for _, err := range []error{treeErr, leftErr, rghtErr} {
+		if err != nil {
+			return nil, nil, nil, err
+		}
+	}
+	return treeOut, leftOut, rghtOut, nil
+}
+
+func (ctrl *Controller) loadTree(gi *storage.GetInput, startTime, endTime time.Time) (_ *storage.GetOutput, _err error) {
+	defer func() {
+		rerr := recover()
+		if rerr != nil {
+			_err = fmt.Errorf("panic: %v", rerr)
+		}
+	}()
```

Comment:
Consider implementing `recoveryMiddleware` that would wrap all handlers and catch all panics.

Reply:
Actually, this method is intended to be called in `loadTreeConcurrently`:

```go
var treeErr, leftErr, rghtErr error
var wg sync.WaitGroup
wg.Add(3)
go func() { defer wg.Done(); treeOut, treeErr = ctrl.loadTree(gi, treeStartTime, treeEndTime) }()
go func() { defer wg.Done(); leftOut, leftErr = ctrl.loadTree(gi, leftStartTime, leftEndTime) }()
go func() { defer wg.Done(); rghtOut, rghtErr = ctrl.loadTree(gi, rghtStartTime, rghtEndTime) }()
wg.Wait()
```

So middleware is not an option here. We must catch all panics in each goroutine or the whole process will stop.

Reply:
`recoveryMiddleware` is not supposed to be a replacement for this recovery. There are many other places in handlers routines that may cause panic.

Reply:
I'm not sure I understand your last comment. Should I keep that code or use something else?

Reply:
If you are concerned with possible panics, in addition to this panic recovery (which scope is `loadTree`) consider implementing recovery middleware like this: https://github.com/pyroscope-io/pyroscope/blob/3bca77ab0930ef02021719b463a2009a31085450/pkg/server/controller.go#L303
that would catch all panics in all handlers (but, of course, not in goroutines created by them). Although a panic in an HTTP handler does not cause app crash (directly), recovering improves user experience, especially if server logs are collected and parsed.

Regarding panic recovery in general: it may make sense to capture stack trace, otherwise error message may be not informative.

Reply:
No, I mean the code is intended to be used in goroutines, not from an HTTP handler:

```go
func (ctrl *Controller) loadTreeConcurrently(...) {
  var treeErr, leftErr, rghtErr error
  var wg sync.WaitGroup
  wg.Add(3)
	                                       // call here ⤵
  go func() { defer wg.Done(); treeOut, treeErr = ctrl.loadTree(gi, treeStartTime, treeEndTime) }()
  go func() { defer wg.Done(); leftOut, leftErr = ctrl.loadTree(gi, leftStartTime, leftEndTime) }()
  go func() { defer wg.Done(); rghtOut, rghtErr = ctrl.loadTree(gi, rghtStartTime, rghtEndTime) }()
  wg.Wait()
  // ...
}

func (ctrl *Controller) loadTree(...) {
  defer func() { 
    rerr := recover() // <- capture here
    // ...
  }()
  // ...
}
```

We should always capture panic inside a goroutine, or the process will crash. This is different from HTTP handlers.

(for HTTP handlers, the project should already have recover middleware somewhere, it's out of scope for this PR)

Reply:
Okay, thanks @olvrng.

Meanwhile, please take into account:
> Regarding panic recovery in general: it may make sense to capture stack trace, otherwise error message may be not informative.

Reply:
Yes, I see. A stack trace log should be good enough for now. Is it ok?

Reply:
Yep, `debug.Stack()` should tell us everything.